File: Stacks/StacksWithoutLibrary.cpp
#include <iostream>
#include <chrono>
#include <iomanip>

/**
 * Node structure for linked list implementation of stack.
 */
struct Node {
    int data;
    Node* next;

    Node(int value) : data(value), next(nullptr) {}
};

/**
 * Stack implementation using linked list (without STL).
 * Provides O(1) push and pop operations.
 */
class Stack {
private:
    Node* top;
    size_t count;

public:
    Stack() : top(nullptr), count(0) {}

    ~Stack() {
        while (!isEmpty()) {
            pop();
        }
    }

    /**
     * Add element to the top of the stack. O(1) operation.
     */
    void push(int value) {
        Node* newNode = new Node(value);
        newNode->next = top;
        top = newNode;
        count++;
    }

    /**
     * Remove and return top element from the stack. O(1) operation.
     */
    int pop() {
        if (isEmpty()) {
            std::cerr << "Error: Attempting to pop from an empty stack." << std::endl;
            return -1;
        }

        int value = top->data;
        Node* temp = top;
        top = top->next;
        delete temp;
        count--;
        return value;
    }

    /**
     * Check if stack is empty. O(1) operation.
     */
    bool isEmpty() const {
        return top == nullptr;
    }

    /**
     * Return top element without removing it. O(1) operation.
     */
    int peek() const {
        if (isEmpty()) {
            std::cerr << "Error: Stack is empty." << std::endl;
            return -1;
        }
        return top->data;
    }

    /**
     * Return number of elements in stack. O(1) operation.
     */
    size_t size() const {
        return count;
    }
};

/**
 * Demonstrates stack operations using custom linked list implementation.
 * Pushes 5 integers and pops them to show LIFO ordering.
 */
int main() {
    auto startTime = std::chrono::high_resolution_clock::now();
    
    // Create stack instance
    Stack myStack;

    // Push elements
    myStack.push(10);
    myStack.push(20);
    myStack.push(30);
    myStack.push(40);
    myStack.push(50);
    
    // Calculate memory (5 nodes + overhead)
    size_t memoryUsed = sizeof(Stack) + (sizeof(Node) * 5);

    // Pop and print elements
    std::cout << "Elements in the Stack:" << std::endl;
    while (!myStack.isEmpty()) {
        std::cout << "Popped Element: " << myStack.pop() << std::endl;
    }
    
    auto endTime = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(endTime - startTime);
    
    // Performance statistics
    std::cout << "\n--- Performance Statistics ---" << std::endl;
    std::cout << std::fixed << std::setprecision(4);
    std::cout << "Execution time: " << duration.count() / 1000.0 << " ms" << std::endl;
    std::cout << "Memory usage: " << memoryUsed << " bytes" << std::endl;
    std::cout << "\nTime Complexity:" << std::endl;
    std::cout << "  - Push: O(1)" << std::endl;
    std::cout << "  - Pop: O(1)" << std::endl;
    std::cout << "  - Peek: O(1)" << std::endl;
    std::cout << "  - IsEmpty: O(1)" << std::endl;

    return 0;
}

File: Stacks/StacksWithoutLibrary.java
package Stacks;

/**
 * Demonstrates stack operations using custom ArrayList implementation.
 * Provides O(1) push and pop operations.
 * Pushes 5 integers and pops them to show LIFO ordering.
 */
public class StacksWithoutLibrary {

    /**
     * Stack implementation using ArrayList.
     */
    private static class Stack {
        private java.util.ArrayList<Integer> items;
        private int count;

        public Stack() {
            this.items = new java.util.ArrayList<>();
            this.count = 0;
        }

        /**
         * Add element to the top of the stack. O(1) operation.
         */
        public void push(int item) {
            items.add(item);
            count++;
        }

        /**
         * Remove and return top element from the stack. O(1) operation.
         */
        public int pop() {
            if (isEmpty()) {
                System.err.println("Error: Attempting to pop from an empty stack.");
                return -1;
            }
            int item = items.get(items.size() - 1);
            items.remove(items.size() - 1);
            count--;
            return item;
        }

        /**
         * Check if stack is empty. O(1) operation.
         */
        public boolean isEmpty() {
            return items.isEmpty();
        }

        /**
         * Return top element without removing it. O(1) operation.
         */
        public int peek() {
            if (isEmpty()) {
                System.err.println("Error: Stack is empty.");
                return -1;
            }
            return items.get(items.size() - 1);
        }

        /**
         * Return number of elements in stack. O(1) operation.
         */
        public int size() {
            return count;
        }
    }

    public static void main(String[] args) {
        Runtime runtime = Runtime.getRuntime();
        runtime.gc();
        long memoryBefore = runtime.totalMemory() - runtime.freeMemory();
        long startTime = System.nanoTime();
        
        // Create stack instance
        Stack myStack = new Stack();

        // Push elements
        myStack.push(10);
        myStack.push(20);
        myStack.push(30);
        myStack.push(40);
        myStack.push(50);

        // Pop and print elements
        System.out.println("Elements in the Stack:");
        while (!myStack.isEmpty()) {
            System.out.println("Popped Element: " + myStack.pop());
        }
        
        long endTime = System.nanoTime();
        long memoryAfter = runtime.totalMemory() - runtime.freeMemory();
        long memoryUsed = memoryAfter - memoryBefore;
        
        // Performance statistics
        System.out.println("\n--- Performance Statistics ---");
        System.out.printf("Execution time: %.4f ms%n", (endTime - startTime) / 1_000_000.0);
        System.out.println("Memory usage: " + (memoryUsed > 0 ? memoryUsed : "negligible") + " bytes");
        System.out.println("\nTime Complexity:");
        System.out.println("  - Push: O(1)");
        System.out.println("  - Pop: O(1)");
        System.out.println("  - Peek: O(1)");
        System.out.println("  - IsEmpty: O(1)");
    }
}
